src/sentry/testutils/helpers/faux.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import absolute_import
+
+from collections import deque
+
+
+class Faux(object):
+    """
+    Convenience functions for testing, and asserting, with ``unittest.mock``
+    objects.
+
+    Usage:
+        >>> with patch('module.ClassName.func') as mock:
+        >>>     ClassName.func('foo', extra={'bar': {'baz': 1}})
+        >>>
+        >>>     assert faux(mock).args == ('foo',)
+        >>>     assert faux(mock).args_contain('foo')
+        >>>     assert faux(mock).kwargs_contain('extra')
+        >>>     assert faux(mock).kwargs_contain('extra.bar')
+        >>>     assert faux(mock).kwarg_equals('extra.bar.baz', 1)
+
+    Dot Notation:
+        Functions that deal with ``kwargs`` will accept a dot notation
+        shorthand for dealing with deeply nested ``dict`` keys.
+
+        >>> with patch('module.ClassName.func') as mock:
+        >>>     ClassName.func('foo', extra={'bar': {'baz': 1}})
+        >>>
+        >>>     assert faux(mock).kwargs_contain('extra.foo.baz')
+        >>>     assert faux(mock).kwarg_equals('extra.foo.baz', 1)
+
+    Multiple Calls
+        By default, ``faux`` will select the last call to a ``mock``. You can
+        specify which call you want by passing the index as a second parameter.
+
+        >>> with patch('module.ClassName.func') as mock:
+        >>>     ClassName.func('foo', extra={'bar': {'baz': 1}})
+        >>>     ClassName.func(None)
+        >>>
+        >>>     assert faux(mock).args == (None,)
+        >>>     assert faux(mock, 0).args == ('foo',)
+    """
+
+    def __init__(self, call):
+        self.call = call
+
+    @property
+    def args(self):
+        return self.call[1]
+
+    @property
+    def kwargs(self):
+        return self.call[2]
+
+    def called_with(self, *args, **kwargs):
+        return self.args == tuple(args) and self.kwargs == kwargs
+
+    def kwargs_contain(self, key):
+        return self._kwarg_exists(key)
+
+    def kwarg_equals(self, key, expected):
+        return self._kwarg_value(key) == expected
+
+    def args_contain(self, value):
+        assert value in self.args
+
+    def args_equals(self, *args):
+        return self.args == tuple(args)
+
+    def _kwarg_exists(self, key):
+        try:
+            self._kwarg_value(key)
+            return True
+        except (KeyError, TypeError):
+            return False
+
+    def _kwarg_value(self, key):
+        """
+        Support a dot notation shortcut for deeply nested dicts or just look
+        up the value if passed a normal key.
+
+        >>> self.kwargs = {'foo': {'bar': {'baz': 1}}}
+        >>> self._kwarg_value('foo.bar.baz')
+        1
+        >>> self._kwarg_value('foo')
+        {'bar': {'baz': 1}}
+        """
+        if '.' in key:
+            keys = deque(key.split('.'))
+        else:
+            return self.kwargs[key]
+
+        kwarg = dict(self.kwargs)
+
+        while keys:
+            kwarg = kwarg[keys.popleft()]
```

Comment:
When there is a KeyError do you want to raise a more verbose assertion error that gives the path that failed? and the data present when the key was missing?

Reply:
Yea, I think it'd be a good idea to add better messages for all failures here. Honestly, I need to read up on `unittest` because my initial (read: naive) attempts to do this didn't change the test output at all.

Reply:
You could catch the KeyError and raise an AssertionError instead.